### DISCUSSION TITLE
Comparison to python-hyper/hyperlink

### DISCUSSION THREAD

**Discussion by ariebovenberg** [aio-libs/yarl#707] [Ideas]
👋  Hi!

I really like that you have a "Comparison with other URL libraries" section in the Readme. Could you perhaps also add https://github.com/python-hyper/hyperlink to the list? I'm curious what the differences are.

**Comment by TayfurYldz** [comment]
A useful high-level comparison is that the two libraries overlap on immutable URL manipulation, but optimize for somewhat different APIs/ecosystems.

yarl.URL is the URL type used heavily in aiohttp/aio-libs. It has first-class helpers for query/path mutation (with_query, update_query, extend_query, with_path, / operator), IDNA/quoting behavior integrated into construction, and cached parsed properties via propcache. It also deliberately exposes user/password/host/port pieces and canonicalizes according to yarl's URL model.

hyperlink's URL/DecodedURL API is more explicitly split between encoded and decoded representations. Its API tends to make percent-encoding/decoding boundaries visible through URL versus DecodedURL, and it has different mutation vocabulary (replace, child, add/set/remove query parameters).

That means I would not choose between them from a feature checklist alone. The important compatibility questions are:

- whether your surrounding stack already uses yarl.URL (aiohttp is the obvious case);
- whether you want a distinct decoded-URL abstraction;
- exact normalization/equality/percent-encoding semantics your application depends on;
- query duplicate/order behavior;
- URI vs IRI handling and IDNA behavior.

They are not drop-in replacements. In particular, code that serializes a URL and later compares it should test its own edge cases (Unicode host/path, already-percent-encoded input, duplicate query keys, empty query/fragment, IPv6/userinfo) before migrating.

For this repository's README comparison, those semantic differences would be more useful than a simple "supports feature X" row, because both projects cover the common URL-building surface fairly well.
